src/path.ts

```diff
@@ -0,0 +1,109 @@
+/* eslint-disable unicorn/prefer-at */
+export type QueryValue =
+  | string
+  | number
+  | boolean
+  | QueryValue[]
+  | Record<string, any>
+  | null
+  | undefined;
+export type QueryObject = Record<string, QueryValue | QueryValue[]>;
+
+/**
+ * Removes the leading slash from the given path if it has one.
+ */
+export function withoutLeadingSlash(path?: string): string {
+  if (!path || path === "/") {
+    return "/";
+  }
+
+  return path[0] === "/" ? path.slice(1) : path;
+}
+
+/**
+ * Removes the trailing slash from the given path if it has one.
+ */
+export function withoutTrailingSlash(path?: string): string {
+  if (!path || path === "/") {
+    return "/";
+  }
+
+  return path[path.length - 1] === "/" ? path.slice(0, -1) : path;
+}
+
+/**
+ * Joins the given base URL and path, ensuring that there is only one slash between them.
+ */
+export function joinURL(base?: string, path?: string): string {
+  if (!base || base === "/") {
+    return path || "/";
+  }
+
+  if (!path || path === "/") {
+    return base || "/";
+  }
+
+  const baseHasTrailing = base[base.length - 1] === "/";
+  const pathHasLeading = path[0] === "/";
+  if (baseHasTrailing && pathHasLeading) {
+    return base + path.slice(1);
+  }
+
+  if (!baseHasTrailing && !pathHasLeading) {
+    return `${base}/${path}`;
+  }
+
+  return base + path;
+}
+
+/**
+ * Adds the base path to the input path, if it is not already present.
+ */
+export function withBase(input = "", base = ""): string {
+  if (!base || base === "/") {
+    return input;
+  }
+
+  const _base = withoutTrailingSlash(base);
+  if (input.startsWith(_base)) {
+    return input;
+  }
+
+  return joinURL(_base, input);
+}
+
+/**
+ * Returns the URL with the given query parameters. If a query parameter is undefined, it is omitted.
+ */
+export function withQuery(input: string, query: QueryObject): string {
+  if (!query || Object.keys(query).length === 0) {
+    return input;
+  }
+
+  const searchIndex = input.indexOf("?");
+  const base = searchIndex === -1 ? input : input.slice(0, searchIndex);
+  const searchParams = new URLSearchParams(
+    searchIndex === -1 ? "" : input.slice(searchIndex + 1)
```

Comment:
fast-path: when there is no searchIndex we can avoid creating new `URLSearchParams` (sorry for being strict this is kind of thing we forget later and it can be potentially bases to `ufo/lite` trying to make sure optimizing it as much as we can)

Reply:
Sorry, I don't get it completely: Even if no search is in the input path, we still need to construct the encoded query string for the new output path, am I right?
How should we build the search params if not using `URLSearchParams`? Port the `stringifyQuery` function from `ufo`? 

P.S.: Of course, no need to justify optimization and future-proofness. 🙋‍♂️

Reply:
Sorry my wording was confusing. I meant that when there is no quesry in input string, we can have a fast path if block. So we don't need to iterate to for example remove `undefined` values, etc (we already know query state is empty) so it can be something like `return new URLSearchParams(query).toString` (I'm not sure if query object needs normalization but it could be simple map)

Reply:
Ahh, I get it now! Thanks for the clarification. Now:
1. If the query is empty, we return early.
2. If the input URL has no query, we only normalize the items of the input `query` object.

We have to normalize the values for `URLSearchParams`, otherwise it returns string like `key=undefined`.

I have again tested with the (slighty changed) `ufo` tests. I will create a separate PR there to track the implementation of these lighter functions.